sdk/src/main/java/com/eegeo/mapapi/widgets/RouteViewHelper.java

```diff
@@ -0,0 +1,159 @@
+package com.eegeo.mapapi.widgets;
+
+import android.location.Location;
+import android.util.Pair;
+
+import com.eegeo.mapapi.geometry.LatLng;
+import com.eegeo.mapapi.services.routing.RouteStep;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+class RouteViewHelper {
+
+    public static boolean areApproximatelyEqual(LatLng firstLocation, LatLng secondLocation) {
+        final double epsilonSq = 1e-6;
+        float[] results = new float[1];
+        Location.distanceBetween(firstLocation.latitude, firstLocation.longitude, secondLocation.latitude, secondLocation.longitude, results);
+        return results[0] <= epsilonSq;
+    }
+
+    public static boolean areCoordinateElevationPairApproximatelyEqual(Pair<LatLng, Double> a, Pair<LatLng, Double> b) {
+        final double elevationEpsilon = 1e-3;
+
+        if(!areApproximatelyEqual(a.first, b.first)) {
+            return false;
+        }
+
+        return Math.abs(a.second - b.second) < elevationEpsilon;
+    }
+
+    public static void removeCoincidentPoints(List<LatLng> coordinates) {
+        if(coordinates == null) {
+            return;
+        }
+
+        Iterator<LatLng> it = coordinates.iterator();
+        LatLng prev = null;
+        LatLng current = it.next();
+        while (it.hasNext()) {
+            prev = current;
+            current = it.next();
+            if (areApproximatelyEqual(current, prev)) {
+                it.remove();
+            }
+        }
+    }
+
+    public static void removeCoincidentPointsWithElevations(List<LatLng> coordinates, List<Double> perPointElevations) {
+        List<Pair<LatLng, Double>> pairsList = new ArrayList<>(coordinates.size());
+
+        for(int i=0; i<coordinates.size(); i++) {
+            LatLng coordinate = coordinates.get(i);
+            Double elevation = perPointElevations.get(i);
+
+            pairsList.add(new Pair<>(coordinate, elevation));
+        }
+
+        int i = 1;
+        while (i < pairsList.size()) {
+            Pair<LatLng, Double> prev = pairsList.get(i-1);
+            Pair<LatLng, Double> current = pairsList.get(i);
+
+            if (areCoordinateElevationPairApproximatelyEqual(current, prev)) {
+               pairsList.remove(i-1);
+               coordinates.remove(i-1);
+               perPointElevations.remove(i-1);
+            } else {
+                i++;
+            }
```

Comment:
We are iterating on a list while also removing its elements. Suggest using a different approach. Maybe there is a java alternative for `std:unique` or we could use iterators like in `removeCoincidentPoints`

Reply:
chnaged